### PR TITLE
[Autoscaler] Reload config 

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This ensures that available node types is reloaded correctly, which involves moving the loading logic into reload and updating the launch hashes accordingly. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/tests/test_autoscaler.py [hunk 1]
```diff
@@ -105,24 +105,29 @@ def __init__(self, cache_stopped=False):
         self.ready_to_create = threading.Event()
         self.ready_to_create.set()
         self.cache_stopped = cache_stopped
+        # Many of these functions are called by node_launcher or updater in
+        # different threads. This can be treated as a global lock for everything.
+        self.lock = threading.Lock()
```
Comment (wuisawesome) [inline python/ray/tests/test_autoscaler.py:110]: In practice, this really reduces the amount of race conditions _especially_ when print statements/io are used.

File: python/ray/autoscaler/autoscaler.py [hunk 2]
```diff
@@ -60,18 +60,6 @@ def __init__(self,
         self.config_path = config_path
         self.reload_config(errors_fatal=True)
```
Comment (ericl) [inline python/ray/autoscaler/autoscaler.py:61]: Can we rename reload_config to `reset()` now that's it doing more stuff?

File: python/ray/autoscaler/node_launcher.py [hunk 3]
```diff
@@ -36,7 +36,10 @@ def _launch_node(self, config: Dict[str, Any], count: int,
             assert node_type, node_type
         worker_filter = {TAG_RAY_NODE_KIND: NODE_KIND_WORKER}
         before = self.provider.non_terminated_nodes(tag_filters=worker_filter)
-        launch_hash = hash_launch_conf(config["worker_nodes"], config["auth"])
+        launch_hash = hash_launch_conf(
+            config["worker_nodes"],
+            config["auth"],
+            available_node_types=config.get("available_node_types"))
```
Comment (ericl) [inline python/ray/autoscaler/node_launcher.py:42]: We shouldn't include the entire config as part of the launch conf, or you need to relaunch if any config changes. Can you extract just the relevant one?

Comment (ericl) [reply python/ray/autoscaler/node_launcher.py:42]: Same comment for launch_hash in commands.py

**Review by ericl — CHANGES_REQUESTED** [review]

**Comment by ericl** [comment]
Can you also add unit tests covering the new reload behavior?

**Review by ericl — APPROVED** [review]

**Pull request merged**